### ISSUE TITLE
Possibly unbound symbols are assumed to be bound in certain cases

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#5548]
**Describe the bug**

Pyright correctly infers a conditionally import symbol could be `Unbound` in one part of the program. However later in the program it incorrectly makes the assumption that it is not. This decision seems to change if the symbol is used elsewhere in a certain way.

**To Reproduce**

```python
import os

if os.getenv("TEST"):
    from itertools import islice
else:
    pass

# islice([1,2,3], 2) # Ex 1

test0 = lambda: islice([1,2,3], 2) # Ex 2

def test1():
    return islice([1,2,3], 2) # Ex 3
```

When Ex 1 is commented, pyright incorrectly reports `islice` is bound in both Ex 2 & 3. Uncommenting Ex 2 causes pyright to work correctly and report that `islice` may be unbound.

Perhaps a separate issue, but the hover tip for `islice` always says `class islice` and doesn't indicate if it is possibly unbound, even when pyright reports a type error stating that it is. This seems to only happen with classes, because changing `islice` to `tee` does show `... | Unbound`.

**Expected behavior**

Pyright should report that `islice` is potentially unbound, even if I comment out Ex 1.

**VS Code extension or command-line**

I'm using pyright v1.1.316 as a language server with neovim.

**Comment by erictraut** [comment]
This is by design. Detection of potentially-unbounded variables is performed within an execution scopes — i.e. within a code flow graph. It is not performed across execution scopes because a static type checker cannot generally determine when different execution scopes will be run relative to each other. In your example, `isslice` is defined in the module execution scope but is accessed from within the lambda or a function scope.

If pyright were to report a "potentially unbound symbol" here, it would also produce false positives in many other common situations.

**Comment by kyle-query** [comment]
> In your example, isslice is defined in the module execution scope but is accessed from within the lambda or a function scope.

Thanks, that does makes sense as uncommenting Ex 2 means `isort` is both defined *and* accessed within a module scope, so we should see an error at Ex 2.

However uncommenting Ex 2 *also* results in an error at Ex 1 & 3. Why does uncommenting it cause pyright to behave like it is checking across execution scopes in that case?

**Comment by erictraut** [comment]
Ah, I missed the original point about "uncommenting Ex 2". That does look like an inconsistency. I'll reopen the issue and investigate further.

**Comment by erictraut** [comment]
I found the bug that result in the apparent inconsistency. It was a subtle interaction between unbound symbol detection and a feature that allows type narrowing of variables captured by inner scopes under certain circumstances. Normally, code flow analysis isn't performed across execution scope boundaries, but it is in the specific case where a captured variable is used within an inner scope and pyright is able to prove the variable is not modified along any code path after it is captured. That applies to your code sample above. Unbound variable detection is also deterministically safe (no risk of false positives) under the same set of conditions.

I've fixed the bug so unbound variable detection works for captured variables when it can be safely applied. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
